File: app/src/main/java/com/weathercock/cleanarchitecture/data/dao/CityDao.java
package com.weathercock.cleanarchitecture.data.dao;

import android.content.Context;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.weathercock.cleanarchitecture.data.entity.city.PrefEntity;

import org.json.JSONException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.List;

/**
 * City Data Access Object.
 * <p>
 * Created by weath on 2016/06/14.
 */
public class CityDao {

    private Context context;

    public CityDao(Context context) {
        this.context = context;
    }

    public List<PrefEntity> find() throws IOException, JSONException {
        String cityJson = readJson();
        return parseJson(cityJson);
    }

    private String readJson() throws IOException {
        InputStream is = null;
        BufferedReader br = null;
        String text = "";

        try {
            is = context.getAssets().open("json/city.json");
            br = new BufferedReader(new InputStreamReader(is));

            String str;
            while ((str = br.readLine()) != null) {
                text += str + "\n";
            }
        } finally {
            if (is != null) is.close();
            if (br != null) br.close();
        }

        return text;
    }

    private List<PrefEntity> parseJson(String jsonStr) throws JSONException {
        Gson gson = new Gson();
        List<PrefEntity> entityList = gson.fromJson(jsonStr, new TypeToken<List<PrefEntity>>() {
        }.getType());

        return entityList;
    }

}

File: app/src/main/java/com/kazakago/cleanarchitecture/data/dao/CityDao.kt
package com.kazakago.cleanarchitecture.data.dao

import android.content.Context
import com.kazakago.cleanarchitecture.data.entity.city.PrefEntity
import com.squareup.moshi.KotlinJsonAdapterFactory
import com.squareup.moshi.Moshi
import com.squareup.moshi.Types
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader

class CityDao(private val context: Context) {

    @Throws(IOException::class)
    fun find(): List<PrefEntity> {
        val cityJson = readJson()
        return parseJson(jsonStr = cityJson)
    }

    @Throws(IOException::class)
    private fun readJson(): String {
        var text = ""
        context.assets.open("json/city.json").use {
            BufferedReader(InputStreamReader(it)).use {
                var str = it.readLine()
                while (str != null) {
                    text += str + "\n"
                    str = it.readLine()
                }
            }
        }
        return text
    }

    private fun parseJson(jsonStr: String): List<PrefEntity> {
        val moshi = Moshi.Builder()
                .add(KotlinJsonAdapterFactory())
                .build()
        val type = Types.newParameterizedType(List::class.java, PrefEntity::class.java)
        val adapter = moshi.adapter<List<PrefEntity>>(type)
        return adapter.fromJson(jsonStr).orEmpty()
    }

}
